src/crayipxe/tokens.py

```diff
@@ -0,0 +1,136 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import jwt
+import time
+import logging
+
+import oauthlib.oauth2
+import requests_oauthlib
+
+LOGGER = logging.getLogger(__name__)
+TOKEN_HOST = "api-gw-service-nmn.local"  # default in case it is not in the settings configmap
+
+def fetch_token(token_host=TOKEN_HOST):
+    # The token will be fetched from Keycloak using the client id and secret
+    # from the mounted Kubernetes secret.
+    token_url = "https://%s/keycloak/realms/shasta/protocol/openid-connect/token" % token_host
+    auth_user_file = "/client_auth/client-id"
+    auth_secret_file = "/client_auth/client-secret"
+    oauth_client_id = ""
+    oauth_client_secret = ""
+    f = None
+    try:
+        f = open(auth_user_file, 'r')
+        oauth_client_id = f.readline().rstrip()
+    except IOError:
+        LOGGER.error("Unable to read user name from %s", auth_user_file)
+        return None
+    finally:
+        if f:
+            f.close()
+            f = None
+    try:
+        f = open(auth_secret_file, 'r')
+        oauth_client_secret = f.readline().rstrip()
+    except IOError:
+        LOGGER.error("Unable to read secret from %s", auth_secret_file)
+        return None
+    finally:
+        if f:
+            f.close()
+            f = None
+
+    oauth_client = oauthlib.oauth2.BackendApplicationClient(
+        client_id=oauth_client_id)
+
+    session = requests_oauthlib.OAuth2Session(
+        client=oauth_client, auto_refresh_url=token_url,
+        auto_refresh_kwargs={
+            'client_id': oauth_client_id,
+            'client_secret': oauth_client_secret,
+        },
+        token_updater=lambda t: None)
+
+    # Set the CA Cert file location so that we can use TLS to talk with Keycloak.
+    # This certificate is mounted from an existing configmap.
+    session.verify = "/ca_public_key/certificate_authority.crt"
+
+    token = session.fetch_token(token_url=token_url, client_id=oauth_client_id,
+                                client_secret=oauth_client_secret, timeout=500)
+
+    access_token = None
+    if token:
+        access_token = token.get("access_token")
+        if (access_token is not None):
+            LOGGER.debug("Got access_token %s", access_token)
+            return access_token
+        else:
+            LOGGER.error("Unable to get an access_token for client %s",
+                         oauth_client_id)
+            return None
+    else:
+        LOGGER.error("Unable to get a token object for client %s",
+                     oauth_client_id)
+        return None
+
```

Comment:
Offering style suggesting (not required)
```suggestion

    access_token = None
    if not token:
        LOGGER.error("Unable to get a token object for client %s",
                     oauth_client_id)
        return None

    access_token = token.get("access_token")
    if not access_token:
        LOGGER.error("Unable to get an access_token for client %s",
                        oauth_client_id)
        return None

    LOGGER.debug("Got access_token %s", access_token)
    return access_token

```